src/lib.rs

```diff
@@ -71,7 +71,46 @@ impl BigInt<[Digit; 1]> {
     }
 }
 
-impl<D1, D2> std::ops::Add<&BigInt<D2>> for &BigInt<D1>
+macro_rules! impl_digit_operation {
+    ($name:ident,$op:ident,$tp:ty) => {
+        impl<D: AsRef<[Digit]>> $name<$tp> for &BigInt<D> {
+            type Output = BigInt;
+
+            #[inline(always)]
+            fn $op(self, rhs: $tp) -> Self::Output {
+                self.$op(Digit::from(rhs))
+            }
+        }
+    };
+}
+
+macro_rules! impl_digit_operation_reverse {
+    ($name:ident,$op:ident,$tp:ty) => {
+        impl<'n, D: AsRef<[Digit]>> $name<&'n BigInt<D>> for $tp {
+            type Output = BigInt;
+
+            #[inline(always)]
+            fn $op(self, rhs: &BigInt<D>) -> Self::Output {
+                rhs.$op(self)
+            }
+        }
+    };
+}
+
+macro_rules! impl_u64_operation {
+    ($name:ident,$op:ident) => {
+        impl<D: AsRef<[Digit]>> $name<u64> for &BigInt<D> {
+            type Output = BigInt;
+
+            #[inline(always)]
+            fn $op(self, rhs: u64) -> Self::Output {
+                self.$op(&BigInt::from_digits(&U64::from(rhs)))
```

Comment:
Possible optimisation:

```suggestion
                match &*U64::from(rhs) {
                    &[digit] => self.$op(digit),
                    digits => self.$op(&BigInt::from_digits(digits)),
                }
```

The point is to use cheaper addition with digit on x64 platforms. I think the compiler should be able to optimise out the second branch that's never reached on specific platform?

Reply:
Sorry, somehow missed this one. Will address it in the next commit

Reply:
Don't worry, there's no way you could write this lot of code, and take into account every possible optimisation 🙂

Reply:
@survived this macros for `u64` operation was merged with the macros for `u128`. Do you prefer me to separate them again and only implement this optimisation for `u64`?